SourceCode/Homework02/FigureDrawing.cpp

```diff
@@ -0,0 +1,113 @@
+#include <fstream>
+#include <vector>
+#include <functional>
+
+static const unsigned IMAGE_WIDTH = 1920;
+static const unsigned IMAGE_HEIGHT = 1080;
+static const unsigned MAX_COLOR_COMPONENT = 255;
+
+static const char LINE_SEP = '\n';
+static const char PIXEL_SEP = '\t';
+
+struct Color {
+    uint8_t r;
+    uint8_t g;
+    uint8_t b;
+};
+
+typedef std::vector<std::vector<Color>> Image;
+
+void savePixel(std::ofstream& ppm, const Color& pixel);
+void saveImage(const char* imgName, const Image& image);
+void saveImage(std::ofstream& ppm, const Image& image);
+
+
+
+void savePixel(std::ofstream& ppm, const Color& pixel) {
+    // we save control over what to print before and after each pixel
+    ppm << (int)pixel.r << " " << (int)pixel.g << " " << (int)pixel.b; 
+}
+
+void saveImage(const char* imgName, const Image& image) {
+    std::ofstream ppmFileStream(imgName, std::ios::out | std::ios::trunc);
+    ppmFileStream << "P3" << LINE_SEP;
+    ppmFileStream << IMAGE_WIDTH << " " << IMAGE_HEIGHT << LINE_SEP;
+    ppmFileStream << MAX_COLOR_COMPONENT << LINE_SEP;
+
+    saveImage(ppmFileStream, image);
+
+    ppmFileStream.close();
+}
+
+void saveImage(std::ofstream& ppm, const Image& image) {
+    for (int rowId = 0; rowId < IMAGE_HEIGHT; rowId++) {
+        for (int colId = 0; colId < IMAGE_WIDTH; colId++) {
+            savePixel(ppm, image[rowId][colId]);
+            if (colId != IMAGE_WIDTH) {
+                ppm << PIXEL_SEP;
+            }
+        }
+        if (rowId < IMAGE_HEIGHT) { // so we don't save extra lines after the end of the image
+            ppm << LINE_SEP;
+        }
+    }
+}
+
+void createFigure(Image& image, const Color& figureColor, std::function<bool(int, int)> figurePredicate) {
```

Comment:
I personally would prefer going with Shape classes each of which will hold a create function and a predicate, this way messing put the passing of the predicate would be harder.